server/src/main/java/org/elasticsearch/index/mapper/Mapper.java

```diff
@@ -22,6 +24,49 @@
 
 public abstract class Mapper implements ToXContentFragment, Iterable<Mapper> {
 
+    public static final NodeFeature STORE_SOURCE_MAPPER_PARAM = new NodeFeature("mapper.source.store_param");
+
+    static final String STORE_SOURCE_PARAM = "store_source";
+
+    public enum StoreSourceMode {
+        DISABLED("disabled"),  // No source recording
+        ARRAYS("arrays"),      // Store source for arrays of mapped fields
+        ENABLED("enabled");    // Store source for both singletons and arrays of mapped fields
```

Comment:
Not excited about the names here.. Maybe DISABLED => OFF , ENABLED => FULL or ALL?

Reply:
Maybe naive question, but why do we need to enum here? For now at least it seems that just the `index.mapping.store_array_source` setting is sufficient?

Reply:
The idea is that we can add the FULL option for some fields, in case we want to track their source even if they're singletons, e.g. for geo or for objects with deep structure like nested. This option will never be set at the index level as it would disable synthetic source.

Reply:
I think that @jimczi also expressed the desire to generalize the option of storing the source of objects or fields verbatim.

Reply:
Updated to none/arrays/all, ptal.